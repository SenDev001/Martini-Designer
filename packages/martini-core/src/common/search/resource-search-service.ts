export type SearchRecordType = "path";

export interface ResourceSearchResult {
    type: SearchRecordType;
    location: string;
}

export interface FileSearchResult extends ResourceSearchResult {
    directory: boolean;
    name: string;
    extension: string;
    lastModified: number;
}

export namespace FileSearchResult {
    export function is(object: any): object is FileSearchResult {
        return object.type === "path";
    }
}

export const resourceSearchServicePath = "services/martini/resource-search";

export interface SearchQuery {
    query: string;
    type?: SearchRecordType;
    extension?: string;
}

export const ResourceSearchService = Symbol("ResourceSearchService");

export interface ResourceSearchService {
    search(query: SearchQuery): Promise<ResourceSearchResult[]>;
}
