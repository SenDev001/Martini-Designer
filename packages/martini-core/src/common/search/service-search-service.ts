export interface ServiceSearchResult {
    contentType: ContentType;
    location: string;
    name: string;
    description: string;
    qualifiedServiceName: string;
    returnType: ServiceParameter;
    parameters: ServiceParameter[];
    exceptions: string[];
}

export type ContentType = "gloop" | "code" | "bean" | "gtpl";

export interface ServiceParameter {
    typeName: string;
    name: string;
    description: string;
}

export interface SearchQuery {
    query: string;
    contentTypes?: ContentType[];
}

export const serviceSearchServicePath = "services/martini/service-search";

export const ServiceSearchService = Symbol("ServiceSearchService");

export interface ServiceSearchService {
    search(query: SearchQuery): Promise<ServiceSearchResult[]>;
}
