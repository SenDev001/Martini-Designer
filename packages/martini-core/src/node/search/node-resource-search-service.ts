import { ConnectionHandler, JsonRpcConnectionHandler } from "@theia/core";
import { inject, injectable, interfaces } from "inversify";
import { flatten } from "lodash";
import {
    FileSearchResult,
    SearchQuery,
    ResourceSearchResult,
    ResourceSearchService,
    resourceSearchServicePath
} from "../../common/search/resource-search-service";
import { AxiosInstanceFactory } from "../http/axios-instance-factory";

@injectable()
export class ResourceSearchServiceNode implements ResourceSearchService {
    @inject(AxiosInstanceFactory)
    private readonly axiosFactory: AxiosInstanceFactory;

    async search(query: SearchQuery): Promise<ResourceSearchResult[]> {
        const response = await (await this.axiosFactory.make()).get("/coder-api/search", {
            params: {
                q: query.query,
                extension: query.extension,
                type: query.type
            }
        });

        const body = response.data;

        if (response.status !== 200) {
            throw new Error(`Failed to search resource: ${JSON.stringify(body)}`);
        }

        return flatten(
            body
                .map((result: any) => {
                    const baseLocation = result.location || "";
                    if (result.records)
                        return result.records.map((record: any) => {
                            if (record.type === "path")
                                return <FileSearchResult>{
                                    type: "path",
                                    location: "/" + baseLocation + "/" + record.location,
                                    name: record.name,
                                    extension: record.extension,
                                    directory: record.file_type === "directory",
                                    lastModified: record.last_modified
                                };
                        });
                })
                .filter((records: any) => records !== undefined)
        );
    }
}

export const bindResourceSearchService = (bind: interfaces.Bind) => {
           bind(ResourceSearchService)
               .to(ResourceSearchServiceNode)
               .inSingletonScope();
           bind(ConnectionHandler)
               .toDynamicValue(
                   ctx =>
                       new JsonRpcConnectionHandler(resourceSearchServicePath, _ =>
                           ctx.container.get(ResourceSearchService)
                       )
               )
               .inSingletonScope();
       };
