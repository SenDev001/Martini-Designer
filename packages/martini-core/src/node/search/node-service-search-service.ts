import { ConnectionHandler, JsonRpcConnectionHandler } from "@theia/core";
import { inject, injectable, interfaces } from "inversify";
import { flatten } from "lodash";
import {
    SearchQuery,
    ServiceSearchResult,
    ServiceSearchService,
    serviceSearchServicePath
} from "../../common/search/service-search-service";
import { AxiosInstanceFactory } from "../http/axios-instance-factory";

@injectable()
export class ServiceSearchServiceNode implements ServiceSearchService {
    @inject(AxiosInstanceFactory)
    private readonly axiosFactory: AxiosInstanceFactory;

    async search(query: SearchQuery): Promise<ServiceSearchResult[]> {
        const response = await (await this.axiosFactory.make()).get("/coder-api/search/service", {
            params: {
                q: query.query,
                contentType: query.contentTypes
                    ? query.contentTypes.map(type => type.replace("code", "gloop-service")).join(",")
                    : undefined
            }
        });

        const body = response.data;

        if (response.status !== 200) throw new Error(`Failed to search service: ${JSON.stringify(body)}`);

        return flatten(
            body.matches.map((match: any) =>
                match.entries.map((entry: any) => ({
                    contentType: match.content_type === "gloop-service" ? "code" : match.content_type,
                    location: match.location,
                    name: entry.name,
                    description: entry.description,
                    qualifiedServiceName: entry.qualifiedServiceName,
                    returnType: entry.returnType,
                    parameters: entry.parameters,
                    exceptions: entry.exceptions
                }))
            )
        );
    }
}

export const bindServiceSearchService = (bind: interfaces.Bind) => {
    bind(ServiceSearchService)
        .to(ServiceSearchServiceNode)
        .inSingletonScope();
    bind(ConnectionHandler)
        .toDynamicValue(
            ctx => new JsonRpcConnectionHandler(serviceSearchServicePath, _ => ctx.container.get(ServiceSearchService))
        )
        .inSingletonScope();
};
