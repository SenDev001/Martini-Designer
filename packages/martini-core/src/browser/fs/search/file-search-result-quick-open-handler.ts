import { OpenerService, WidgetManager } from "@theia/core/lib/browser";
import { QuickOpenItemOptions, QuickOpenMode } from "@theia/core/lib/common/quick-open-model";
import URI from "@theia/core/lib/common/uri";
import { inject, injectable } from "inversify";
import { Directory, Resource } from "../../../common/fs/martini-filesystem";
import { PartialMartiniPackage } from "../../../common/package/martini-package-manager";
import { FileSearchResult, ResourceSearchResult } from "../../../common/search/resource-search-service";
import { Navigator, NavigatorTreeLabelProvider } from "../../navigator/martini-navigator-view-widget";
import { ResourceSearchResultQuickOpenHandler } from "../../search/resource-search-quick-open";

@injectable()
export class FileSearchResultQuickOpenHandler implements ResourceSearchResultQuickOpenHandler {
    @inject(OpenerService)
    private readonly openerService: OpenerService;
    @inject(NavigatorTreeLabelProvider)
    private readonly labelProvider: NavigatorTreeLabelProvider;
    @inject(WidgetManager)
    private readonly widgetMngr: WidgetManager;

    supports(result: ResourceSearchResult): boolean {
        return FileSearchResult.is(result);
    }

    getOptions(result: FileSearchResult): QuickOpenItemOptions {
        const resource: Resource | Directory = {
            location: result.location,
            name: result.name,
            directory: result.directory,
            lastModified: result.lastModified,
            readOnly: false,
            size: 0,
            children: result.directory ? [] : undefined
        };

        return {
            label: result.location,
            iconClass: this.labelProvider.getIconClass(resource),
            run: mode => {
                if (mode === QuickOpenMode.OPEN || mode === QuickOpenMode.OPEN_IN_BACKGROUND) return this.open(result);
                return false;
            }
        };
    }

    private open(result: ResourceSearchResult): boolean {
        let uri: URI | undefined;
        if (FileSearchResult.is(result)) {
            if (!result.directory) uri = new URI(`martini:/${result.location}`);
            else {
                this.openDir(result);
                return true;
            }
        }
        if (uri) {
            this.openerService.getOpener(uri).then(opener => opener.open(uri!));
            return true;
        }

        return false;
    }

    private async openDir(result: FileSearchResult): Promise<void> {
        const navigator = (await this.widgetMngr.getOrCreateWidget(Navigator.ID)) as Navigator;
        await navigator.model.expandElements(element => {
            if (PartialMartiniPackage.is(element)) return result.location.startsWith(element.name, 1);

            return Resource.is(element) && element.directory && result.location.startsWith(element.location);
        });
        await navigator.model.selectElements(
            element => Resource.is(element) && element.directory && result.location === element.location
        );
        navigator.activate();
    }
}
