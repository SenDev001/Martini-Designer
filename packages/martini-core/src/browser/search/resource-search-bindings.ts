import { QuickOpenContribution } from "@theia/core/lib/browser";
import { interfaces } from "inversify";
import { FileSearchResultQuickOpenHandler } from "../fs/search/file-search-result-quick-open-handler";
import {
    ResourceSearchQuickOpenContribution,
    ResourceSearchQuickOpenHandler,
    ResourceSearchResultQuickOpenHandler
} from "./resource-search-quick-open";

export const bindResourceSearchBindings = (bind: interfaces.Bind) => {
    bind(QuickOpenContribution)
        .to(ResourceSearchQuickOpenContribution)
        .inSingletonScope();
    bind(ResourceSearchQuickOpenHandler)
        .toSelf()
        .inSingletonScope();
    bind(ResourceSearchResultQuickOpenHandler)
        .to(FileSearchResultQuickOpenHandler)
        .inSingletonScope();
};
