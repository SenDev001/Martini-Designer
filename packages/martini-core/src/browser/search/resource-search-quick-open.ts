import {
    QuickOpenContribution,
    QuickOpenHandler,
    QuickOpenHandlerRegistry,
    QuickOpenOptions
} from "@theia/core/lib/browser";
import {
    QuickOpenActionProvider,
    QuickOpenItem,
    QuickOpenItemOptions,
    QuickOpenModel
} from "@theia/core/lib/common/quick-open-model";
import { inject, injectable, multiInject } from "inversify";
import { debounce } from "lodash";
import messages from "martini-messages/lib/messages";
import { ResourceSearchResult, ResourceSearchService } from "../../common/search/resource-search-service";

export const ResourceSearchResultQuickOpenHandler = Symbol("ResourceSearchResultQuickOpenHandler");

export interface ResourceSearchResultQuickOpenHandler {
    supports(result: ResourceSearchResult): boolean;
    getOptions(result: ResourceSearchResult): QuickOpenItemOptions;
}

@injectable()
export class ResourceSearchQuickOpenHandler implements QuickOpenModel, QuickOpenHandler {
    private static readonly DEBOUNCE_TIME = 300;
    readonly prefix = "find ";
    readonly description = messages.resource_search_desc;

    @inject(ResourceSearchService)
    private readonly searchService: ResourceSearchService;
    @multiInject(ResourceSearchResultQuickOpenHandler)
    private readonly handlers: ResourceSearchResultQuickOpenHandler[];

    getModel(): QuickOpenModel {
        return this;
    }

    getOptions(): Partial<QuickOpenOptions.Resolved> {
        return {
            placeholder: messages.resource_search_placeholder
        };
    }

    onType(
        lookFor: string,
        acceptor: (items: QuickOpenItem<QuickOpenItemOptions>[], actionProvider?: QuickOpenActionProvider) => void
    ): void {
        this.doOnType.cancel();
        if (lookFor.trim().length === 0) {
            acceptor([]);
            return;
        }

        this.doOnType(lookFor, acceptor);
    }

    private readonly doOnType = debounce(
        (
            lookFor: string,
            acceptor: (items: QuickOpenItem<QuickOpenItemOptions>[], actionProvider?: QuickOpenActionProvider) => void
        ): void => {
            lookFor = lookFor.toLowerCase();
            this.searchService.search({ query: lookFor }).then(results => {
                const items: QuickOpenItem[] = results
                    .filter((result: any) => result.extension !== "class")
                    .map(result => {
                        const handler = this.handlers.find(handler => handler.supports(result));

                        if (handler) {
                            const options = handler.getOptions(result);
                            if (options.label) {
                                // TODO highlighting should be improved.
                                const index = options.label.toLowerCase().indexOf(lookFor);
                                if (index !== -1)
                                    options.labelHighlights = [
                                        {
                                            start: index,
                                            end: index + lookFor.length
                                        }
                                    ];
                            }
                            return new QuickOpenItem(options);
                        }
                        return undefined;
                    })
                    .filter(item => item)
                    .sort((item1, item2) =>
                        (item1!.getLabel() || "").localeCompare(item2!.getLabel() || "")
                    ) as QuickOpenItem[];
                if (items.length !== 0) acceptor(items);
                else
                    acceptor([
                        new QuickOpenItem({
                            label: messages.no_matches
                        })
                    ]);
            });
        },
        ResourceSearchQuickOpenHandler.DEBOUNCE_TIME
    );
}

@injectable()
export class ResourceSearchQuickOpenContribution implements QuickOpenContribution {
    @inject(ResourceSearchQuickOpenHandler)
    private readonly handler: ResourceSearchQuickOpenHandler;

    registerQuickOpenHandlers(handlers: QuickOpenHandlerRegistry): void {
        handlers.registerHandler(this.handler, true);
        (handlers as any).handlers.delete("...");
    }
}
