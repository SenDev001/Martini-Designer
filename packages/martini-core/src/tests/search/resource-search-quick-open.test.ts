import { Container } from "inversify";
import "reflect-metadata";
import {
    ResourceSearchQuickOpenHandler,
    ResourceSearchResultQuickOpenHandler
} from "../../browser/search/resource-search-quick-open";
import { ResourceSearchResult, ResourceSearchService } from "../../common/search/resource-search-service";
import { ResourceSearchServiceNode } from "../../node/search/node-resource-search-service";

jest.mock("../../node/search/node-resource-search-service");

const container = new Container();
container.bind(ResourceSearchQuickOpenHandler).toSelf();
const mockResourceSearchService = new ResourceSearchServiceNode();
container.bind(ResourceSearchService).toConstantValue(mockResourceSearchService);
const mockHandler: ResourceSearchResultQuickOpenHandler = {
    getOptions: jest.fn((result: ResourceSearchResult) => ({
        label: result.location
    })),
    supports: jest.fn()
};
container.bind(ResourceSearchResultQuickOpenHandler).toConstantValue(mockHandler);
const handler = container.get(ResourceSearchQuickOpenHandler);

beforeEach(() => jest.clearAllMocks());

test("Should return options", () => {
    const options = handler.getOptions();
    expect(options).toMatchSnapshot();
});

test("Should return no results", done => {
    (mockResourceSearchService.search as jest.Mock).mockResolvedValue([]);
    const accept = jest.fn(_items => {
        expect(mockResourceSearchService.search).toBeCalledTimes(1);
        expect(mockResourceSearchService.search).toBeCalledWith({ query: "test" });
        expect(accept).toBeCalledTimes(1);
        expect(accept.mock.calls[0][0]).toMatchSnapshot();
        done();
    });
    handler.onType("test", accept);
});

test("Should return items", done => {
    (mockResourceSearchService.search as jest.Mock).mockResolvedValue([
        {
            type: "path",
            location: "test1"
        },
        {
            type: "path",
            location: "test2"
        }
    ] as ResourceSearchResult[]);
    (mockHandler.supports as jest.Mock).mockReturnValue(true);
    const accept = jest.fn(_items => {
        expect(mockResourceSearchService.search).toBeCalledTimes(1);
        expect(mockResourceSearchService.search).toBeCalledWith({ query: "test" });
        expect(accept).toBeCalledTimes(1);
        expect(accept.mock.calls[0][0]).toMatchSnapshot();
        done();
    });
    handler.onType("test", accept);
});

test("Should not return items", done => {
    (mockResourceSearchService.search as jest.Mock).mockResolvedValue([
        {
            type: "path",
            location: "test1"
        },
        {
            type: "path",
            location: "test2"
        }
    ] as ResourceSearchResult[]);
    (mockHandler.supports as jest.Mock).mockReturnValue(false);
    const accept = jest.fn(_items => {
        expect(mockResourceSearchService.search).toBeCalledTimes(1);
        expect(mockResourceSearchService.search).toBeCalledWith({ query: "test" });
        expect(accept).toBeCalledTimes(1);
        expect(accept.mock.calls[0][0]).toMatchSnapshot();
        done();
    });
    handler.onType("test", accept);
});
