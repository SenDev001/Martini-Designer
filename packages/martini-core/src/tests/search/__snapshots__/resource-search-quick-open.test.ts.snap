// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should not return items 1`] = `
Array [
  QuickOpenItem {
    "options": Object {
      "label": "No matches",
    },
  },
]
`;

exports[`Should return items 1`] = `
Array [
  QuickOpenItem {
    "options": Object {
      "label": "test1",
      "labelHighlights": Array [
        Object {
          "end": 4,
          "start": 0,
        },
      ],
    },
  },
  QuickOpenItem {
    "options": Object {
      "label": "test2",
      "labelHighlights": Array [
        Object {
          "end": 4,
          "start": 0,
        },
      ],
    },
  },
]
`;

exports[`Should return no results 1`] = `
Array [
  QuickOpenItem {
    "options": Object {
      "label": "No matches",
    },
  },
]
`;

exports[`Should return options 1`] = `
Object {
  "placeholder": "Type to search for any resource...",
}
`;
